### ISSUE TITLE
Implementation: Start a discussion forum

### ISSUE THREAD

**Issue by jenlampton** [backdrop-ops/backdropcms.org#250]
If we are going to set up a forum as part of backdropcms.org we'll need to decide how (and where) to do it.  I'd like to eat-our-own dogfood and try to build one using content types, taxonomy and views. If it becomes obvious that we can't do what we need, we can also look at using the forum module, advanced forum module, or similar.

I think the forum site should be a separate website from backdropcms.org. Something like forum.backdropcms.org. This is because 1) people will by definition need to be logged into the forum, but not usually to borg itself. 2) most of the things that are most important to the backdrop communities atm (modules, etc) are on the borg site. I think we can limit our risk by separating the two.

I think we should port, and use, the bakery module to allow single-sign-on between the main borg site an the forum site, and make it as invisible as possible to users that they are on a different site. The separation should be technical only.

## MVP Features

Features:
- Ability to share sessions across sites (bakery module)

Pages for the forum:
- Provide a list of all topics (taxonomy terms)
- Provide a list of all posts in a topic (posts tagged with term)
- Provide a threaded list of all comments on a post 

Node / Question features
- Follow a conversation
- Allow a tags field for users to enter whatever they want

Comment / Answer features
- Mark an answer as the `Accepted answer`
- Mark an answer as `Helpful`
- Mark an answer as `Spam or abusive`

Search
- Start with core search

User account features
- All: add a post into a topic (node)
- All: add comments to a post (comment)
- All: flag everything
- Forum moderator role: people empowered to help & promote

## Nice to have:

Pages:
- FAQ page (featured or most common questions)
- Provide a list of all posts by a user on their profile page.

Node / Question features
- Promote as a FAQ 
- Email people when activity happens on the thread
- @mentions that send an email to said person
- Ability for people to bookmark a particular comment for later

Comment / Answer features
- Votes on answers / Some kind of thumbs-up or facebook like type feature
- Ability to sort replies by # of votes
- Auto-unpublish after X reported flags
- Auto-block user ? if new and spamming, or X reported posts

Search
- Do we need to search across multiple sites? (ack!)

User account features:
- Award points (or badges?) over time (+2 for answer, +1 for comment, +½ for reporting spam -1 for posting spam, tie in to auto block!)
- Allow users to subscribe / unsubscribe from email updates

**Comment by docwilmot** [comment]
Personally, I feel that the forum module nearly does what we want: it builds forum using content types, taxonomy and ... not views. So why don't we build a better forum module? That leverages views and Backdrop properly? Essentially turn Forum into a config for building one using content types, taxonomy and views.

I'm just talking here, dont know how probable this is, but I do think that whatever we end up doing may be something other people will want to do. Making it easily reproducible would be nice.

Having said, yes build one using content types, taxonomy and views. And separate site.

**Comment by jenlampton** [comment]
I love this proposal :)

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by Graham-72** [comment]
How are we going to proceed with this? I would like to help. My way forward would be to install Forum on a test site and then adapt it to look like the Wordpress support forum, with help from @docwilmot where necessary. Would this be useful, or perhaps @docwilmot you have other plans?

**Comment by Gormartsen** [comment]
I noticed that original forum format to discuss is getting outdated. 

also there is another problem - original Drupal 7, include advanced forum functionality doesnot cover regular moderator needs

I was doing a lot of extra work on top of Drupal 6 forum 

There is an example: http://www.drupal.ru/tracker

Basically all communication is spinning around tracker section and node add functionality.

Having user related tracker on user/uid page is useful as well. Specially when you need to make a decision if user is spammer or not.

But also there is a need for time based ban. When somebody broke some rules but it is not worth permanent ban. And you need to inform user that he is banned until specified time by 'custom message'

If you get interested, I can help with converting our code from Drupal.ru website for forum functionality.

What we have right now:
- alternative tracker - news feed focused on recent activity
- abuse module. This one show report status attached to node or comment, allowing other users to see that it has been reported and if there is answer - attached moderator message
- dark matter: alternative to unpublished module. Instead of marking comment or node as unpublished, we mark it dm. It become visible for users who agree to see DM materials. It is very useful against trolls activity. Topic starter is capable to mark any comment as DM and any reply to comment become DM market as well. There is also DM counter in profile. Helps to see how much DM content get produced by specified user.
- thank you module. Simple counter attached to node or comment and allow to show :+ counts. It's pretty much like 'Facebook like' , but it has different influence on auditory. You could express like to trolls activities. For example you could add like to phrase: "go to hell" but add 'thank you' sign has completely different meaning to like.
- text filter to prevent posting 'bad words'


It is interesting task but complex enough. Specially from performance perspective. Usually forum functionality get stressed by social activity and any performance issue become a headache.

**Comment by jenlampton** [comment]
Thanks @Gormartsen, this is part of the reason I don't want to use a forum module, but instead build what we need with views. If we can settle on what that is, we can certainly bundle some default configs into a new backdrop forum module for Backdrop that might be more modern and solve more adminstrator needs.

**Comment by Gormartsen** [comment]
@jenlampton I understand views flexibility, but just want to add 'warning' sign to it.

If we compare VIEW output and raw rendered based on theme function - VIEW will take quite a lot of extra resources. It will be extra important from performance perspective on pages with many nodes and comments.

Caching in this situation is actually creating extra performance issues rather than solve them.

I have a good amount of tuning performance situations, when VIEWS was website performance killer and with enabled views cache - it was killing mysql by bandwidth (like 1-5Mb/sec on simple website).

In situations like this, it was only one solution - put cache into file system instead of database.

**Comment by jenlampton** [comment]
Interesting, it will be a good test to see if we have the same issues with a Backdrop site (I expect we will) and perhaps a fantastic use-case for improving performance for all of Backdrop. I see a future @Gormartsen blog post in the works! :)

**Comment by Gormartsen** [comment]
@jenlampton oh no..... :)

**Comment by docwilmot** [comment]
@jenlampton have a look at https://github.com/backdrop-contrib/brain_forum. I started off thinking of getting Forum module to use Views rather than building its own stuff, but I wondered if this approach isn't better: rather than nodes and comments, use a separate forum entity. 

It gets rid of a primary complaint I have with using nodes, that you lose the breadcrumb trail.

It is however currently very basic. But since it uses views, to make it more complex you just need to build a more complex View.

**Comment by jenlampton** [comment]
What's the difference between a forum entity and a node? What structural element is responsible for the "breadcrumb"?

**Comment by docwilmot** [comment]
> What's the difference between a forum entity and a node?

Not sure TBH, except all the node hooks wouldnt be called. Incidentally, Brain Forum was someone's (Rasmus Werling @Rade333) [bachelor's thesis](https://www.theseus.fi/bitstream/handle/10024/37849/Werling_Rasmus.pdf?sequence=1). So for once I have an academic document to explain everything I can't :smile: 

> What structural element is responsible for the "breadcrumb"?

My statement has to do with the fact that the URL for adding a forum topic is still `node/xx` which isnt in the `forum/category/etc` trail. 

I think though, having worked at this and Forum, that to build a forum from scratch using Views looks like a bit of work.

**Comment by docwilmot** [comment]
The forum listing unstyled

![forum](https://cloud.githubusercontent.com/assets/1042283/15836060/011d9bc2-2c01-11e6-9c1c-03ebace9e810.PNG)

A thread with responses. 

![thread localhost](https://cloud.githubusercontent.com/assets/1042283/15836064/05b37f12-2c01-11e6-85ba-75960d8bb0f8.png)

**Comment by jenlampton** [comment]
I'm evaluating brain_forum and already don't like what I see. On the project page it says: 

> We are at the time being unable to develop Brain Forum further. We had a brain storming session about what the roadmap for the first stable release would be and realized that we are still far away from it and we simply do not have the time to continue development now the version our client ordered is done and we are occupied with other projects.

This group has already bitten off more than they can chew. I'm not interested in picking up where they left off and not being able to support our own software.  I'm going to create a list of features we need (and features we want) and we'll do a "comparison of forum solutions" to see which is the best fit.

**Comment by docwilmot** [comment]
I did "a comparison"; if you want Views forums based on nodes/entities as you've said:

- Brain forum
- [Artesian forum](https://www.drupal.org/project/artesian) (abandoned sandbox)
- [Harmony forum](https://www.drupal.org/project/harmony_core) (actually a distribution)

Building it from scratch using nodes is going to be work. Youre going to need as much code as is currently in Forum anyway, to link nodes, comments, taxonomy, and all the CRUD hooks for the above; essentially what Forum does.

So regardless of what the requirements, youre going to need to write at least a custom module. Or make Forum use Views.

So my thought is, if youre writing custom code, why not abandon the outmoded nodes/comments forum mashup, and do it clean with a forum entity?

Which brings you back to the list above: Brain or Harmony.

I ported Brain because it was easier. So we could think about how an entity based forum would look.

**Comment by jenlampton** [comment]
> why not abandon the outmoded nodes/comments forum mashup, and do it clean with a forum entity?

Well the first question would be "Is there really an advantage to using a forum entity?". Sometimes people want to do things "the drupal 7 way" when there was nothing wrong with "the Drupal 6 way". Could this be another unnecessary layer of abstraction?  We have the all the building-blocks we need already: terms, nodes, and comments. 

Then again, Michelle would know. She's the de facto forum expert for Drupal, and if she wanted to do Artesian forum instead of advanced forum (was that one on your list?) there's probably a good reason why!

> Regardless of the requirements, you're going to need to write at least a custom module

My hope was to write a new contrib module.  "Forum" for backdrop! :)

> Which brings you back to the list above: Brain or Harmony.

Or Artesian! It sounds like Brain has been abandoned at this point too (with no hope for an initial release) but I really don't like the sound of it.  Artesian sounds better, but I don't know what state any of the code is in. I read the dissertation (or whatever it was) about Brain, and was unimpressed.  And the whole time I was thinking, why didn't this work go into Artesian?

**Comment by docwilmot** [comment]
> Or Artesian 

True

**Comment by docwilmot** [comment]
> I'm going to create a list of features we need (and features we want)

Is this different from the requirements in the first post? Not sure if that was always there. Should we wait on you for the next step @jenlampton?

This "start a forum" conversation was the first post I ever made in backdrop, 2 yrs ago, so I'm curious  to see it get off the ground.

**Comment by jenlampton** [comment]
I added the list in the first post, but wanted to check out the wp forum to
see if we should add anything else they have. Sorry, should have added a
comment :)

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by mikemccaffrey** [comment]
Hey folks, I am late to weigh in here, but I'd like to advocate for building the forum out of the standard backdrop pieces of content types, taxonomy, and comments. Backdrop is a framework for building website functionality, and one of the things that you can build is a forum. 

For the forks suggesting a forum module, what additional functionality would it provide other than setting up those components? What exactly is special about a forum that sets it apart from the core functionality of Backdrop, which is to create a website that allows for user-generated content?

If we want to make it easy for people to set up a forum with all of those pieces, why not create an install profile for forums that creates nodes, views, taxonomy terms, and layouts? Then after the install profile runs, folks are free to customize the components that were set up however they see fit.

Makse sense?

**Comment by docwilmot** [comment]
Thanks for chiming in @mikemccaffrey. Here's how I got to this point, step by step:
- I feel that having such major functionality only achievable as an Install profile is certainly not an "easy" choice for those users who may not even know what that is.
- I also think that doing it with the native Backdrop pieces (nodes, views, taxonomy terms, and layouts) is ideal; but that is also certainly not "easy" either. And the chances of doing this with these native bits without having any custom code is very likely close to unreasonable.
- I think having a module which builds a Forum is then the best option.
- The existing Forum module does in fact build Forums with all the bits except Views and Layouts. I think we could repurpose Forum module to use Views and Layouts and achieve what we need. THis is the simplest conclusion so far.
- BUT, if we're going to be re-writing Forum, and considering that several persons in Drupal have noted that Forum is limited in how much it can match the "good" forum software (VB, discourse etc) by the fact that it is **not** actaully a forum, but a mashup of sorts of node comments with some glue.
- Then therefore why not consider building a new version based on the same Backdrop bits (entities, views, taxonomy terms, and layouts - not much difference) while we're at it?

This isnt a conclusion but just an exploration of the options.

**Comment by jenlampton** [comment]
I also disagree with the install profile idea, mostly because people don't even really understand what they are. Additionally, most of the time (like here on backdropcms.org) people want to add a forum to an existing site. If we can do all the things that are done in an install profile (adding terms, content types, & fields) via a module, I think that is a much better way to go.

> I think we could repurpose Forum module to use Views and Layouts and achieve what we need. THis is the simplest conclusion so far.

This.  100% this!

> why not consider building a new version based on the same Backdrop bits (entities, views, taxonomy terms, and layouts - not much difference) while we're at it?

I think before we do too much re-architecting forums we need to find out what exactly failed with Drupal forums. That's still the piece of understanding I'm missing. I think it might be 1) features (`limited in how much it can match the "good" forum software`) and 2) performance. If it's features we could add those with other separate add-on modules. If it's performance, we may have some work to do at scale, and I'm not sure how much we can anticipate that before it happens.

**Comment by docwilmot** [comment]
Please see first attempt at moving Forum to Views. Simply created new views and deleted the existing hook_menu items. Haven't  removed the old table code. 

- Needed to create some new handlers for the forum list. Not comfortabe yet with the `is_container` handler yet. Comments welcome. And will probably need to rename some of the handler labels.
- The original Forum module had hyperlinked container labels in the main Forum list, which loaded a page with the forums in that container. I didnt do that, because I've always found it a bit confusing. If its necessary could probably try to bring that back.
- The container labels are currently unstyled (just plain text); will need a bg color.
- The `Last reply` table column had also the last commenter's name, now just the date/time. Could work on that later if thought needed.

**Comment by Gormartsen** [comment]
@docwilmot container could be replaced by taxonomy tags and instead of container list, there is could be just time line (like http://www.drupal.ru/tracker) 

I personally believe that FB with TW so affected people perception that old style (containers with topics) forum is not really applicable anymore. 

If you go this way, there is no to much to develop.

**Comment by docwilmot** [comment]
Thanks @Gormartsen. I agree we could ditch containers altogether. This step is just to get views though. 

Incidentally I forgot to include the default views. Will update later.

**Comment by docwilmot** [comment]
Updated with default views. Please test.

**Comment by Graham-72** [comment]
I have successfully installed Forum 1.x-2.x together with the necessary Views Term Hierarchy module. Thank you for work so far, but I was immediately struck by the dated appearance which has set me wondering what our aim is for this? Do we have two or three examples of modern forums that we should emulate?

I know very little about this topic but I do feel uneasy about the way contributions are stored as comments. I find the Github arrangement works well but there are times when a particular comment explains something and I want to be able to preserve the comment for future reference, perhaps with some personal bookmark.

I am hoping that when we have it our forum will contain lots of useful 'how to' 'what if' and other information. There are other places for what I would call transient chatter.

**Comment by docwilmot** [comment]
Thanks for trying it, and I'm happy you didnt have any problems (I'm assuming).

> I was immediately struck by the dated appearance which has set me wondering what our aim is for this?

The original Forum was part of Drupal core as you know,, so I think adding features would have been a bit more difficult then. Now, we're only limited by imagination and time. Apart from themeing and prettifying though, what other features would be your top priority?

> I do feel uneasy about the way contributions are stored as comments.

Thats always how Forum worked. There was a brief conversation with @jenlampton above about building an entirely new module based on a dedicated forum entity, but short term this was the easiest solution. See https://github.com/backdrop-contrib/brain_forum

> I am hoping that when we have it our forum will contain lots of useful 'how to' 'what if' and other information.

I am planning that this will be in the Docs rather than Forum.

**Comment by Graham-72** [comment]
In https://github.com/backdrop-ops/backdropcms.org/issues/249 @jenlampton suggested we look at WP support forums, and they look good to me but maybe someone has an even better suggestion?

Surely we can produce a similar layout of topics and posts? I guess we now have the capabilty with layouts and views to choose how we present the things we call and store as comments internally.

While I accept that mature how-to's will be in the handbook, I hope that we will find it useful to share ideas in our support forum?

And will it just be for support or will we have a forum for exchanging ideas on future enhancements of Backdrop?

And will all forums be open to public view? Or should there be a contributors-only one?

**Comment by klonos** [comment]
If we add Flag and/or Voting API (plus a voting solution based on it like Vote Up/Down), then we could have comments marked as "the answer" (one per discussion) and other comments marked as helpful. We could then either theme these to be different then other comments, or have them bubble up to the top of the discussion (something like making the comments "sticky").

I've seen such features in other tech-related forums. Would these features make you feel more confident/comfortable with our forum implementation @Graham-72??

PS (slightly OT): we will need Flag down the road anyways for allowing registered users to mark comments as spam/offensive (same as what was implemented in d.org). I think that in d.org it was Rules + Flag that was doing the spam marking magic work (notify the spam squad to take final action). We can have that be Voting + Rules instead, so that nasties cannot exploit the system (marking legit comments as spam). That would allow us to auto-remove nasty comments say for example if 3+ log-time registered users have marked it as such. ...I'll file a separate issue for that once we have forums online.

**Comment by klonos** [comment]
...I say baby steps. Lets get a simple forum set up and we can add features and style it as we go. It will never be "perfect" but we can improve along the way and perhaps when we feel confident we can add forum back to core in the form of a "Forum" installation profile.

For what it's worth, I like what we have planned so far. I have always been a fan of using core at its full potential in order to build specific solutions (like say e-shops etc). It's what has made projects like Commerce and CiviCRM emerge and evolve to become the popular go-to solutions they are today. Perhaps Forum can evolve to be something like that. ...and along the way, it will reveal any core shortcomings that we can address.

**Comment by Graham-72** [comment]
I am sorry to seem negative about this. It is really good that @docwilmot has it ported so quickly and that it installs without any problem. I know very little about the relative merits of different forums from a users perspective. It just struck me that at the moment the appearance looks very dated, for good reasons no doubt, and we could do a lot better. I would like to see something that matches best current practice, whatever that might be. Is that not possible?

**Comment by klonos** [comment]
I guess that if it is a styling issue, then it's best that @wesruv takes a look at it. ...but I know he must be busy with other things right now.

**Comment by wesruv** [comment]
I should be able to make the time for this if it's a priority?

**Comment by Scoppettuolo** [comment]
Hello.

I have a Forum in Drupal 7, and I can´t convert to BackDrop, compatibility for make this is the first, I think. (I can convert the Content Type, but not the Forum).

I use the Forum module but I think thas is not necessary, because I never use the structure (...Old Style...) that the Forum Module, except little details.

Pages:

Provide a list of all topics - I use Views - Style esoTalk.

Member tools:

Ability for people to create a new forum topic (taxonomy/term, perhaps) - Can be Content Type with the taxonomy for choise.

Moderator tools:

- Need Convert Node, and not only for the Forum. (For example...: Pass the Node Forum to the Node Blog, o pass Node Blog to the Node Forum).
- Split, and Merge Node, like the all forums, and not only for the Forum.

Nice to have:

Provide a list of all posts by a user on their profile page. - I use Views in tab in profile of user.
¿Flag for the rest?

An specific Forum Module is not necessary, I think.

**Comment by laryn** [comment]
It would be neat to have an option to use a Forum in a Q&A format similar to Stack Exchange rather than straight chronological -- so you could sort based on the number of up-votes an answer got and mark one response as the chosen answer.

**Comment by klonos** [comment]
@laryn [way ahead of you](https://github.com/backdrop-ops/backdropcms.org/issues/250#issuecomment-231808645):

> If we add Flag and/or Voting API (plus a voting solution based on it like Vote Up/Down), then we could have comments marked as "the answer" (one per discussion) and other comments marked as "helpful". We could then either theme these to be different than other comments, or have them bubble up to the top of the discussion (something like making the comments "sticky").

**Comment by laryn** [comment]
@klonos haha, oops. I did skim the thread before posting, I promise.

**Comment by docwilmot** [comment]
@laryn I understand that the old style forum is not 'necessary' and most of it can be done with Views, but we would still need to provide an easy method for users to set up a forum. The method you describe is still way to complex for the average new person (not previously worked with Backdrop or Drupal, for example). We still need a quick, simple module.

But good news, I've made a good start: Forum module was converted to use Views. So really Forum just ends up being a big installer for displaying Views.

The 2.x branch here: https://github.com/backdrop-contrib/forum/tree/1.x-2.x

I was hoping @jenlampton could have a look one day to see if it fits the purpose. Still hoping.

**Comment by jenlampton** [comment]
> I was hoping @jenlampton could have a look one day to see if it fits the purpose. Still hoping.

As am I!!! :)

**Comment by laryn** [comment]
@docwilmot I completely agree about needing an easy way for people to set up forums. My suggestion was a setting (either global in forum admin, or on a per-forum basis) to allow the forum to function/sort/display differently to act as a Q&A forum rather than a discussion forum. 

I will try to test out 1.x-2.x on the forum module.

**Comment by Scoppettuolo** [comment]
I use this how Forum...:

http://i.imgur.com/AsigDj9.png
http://i.imgur.com/x0XVgyq.png

**Comment by docwilmot** [comment]
@jenlampton see forum at http://1591.backdrop.backdrop.qa.backdropcms.org/forum. I actually pushed it as a PR to core just to get a sandbox. And see code at https://github.com/backdrop/backdrop/pull/1591/files.

We're having so many discussions which would be best served in a central forum its about time.

**Comment by jenlampton** [comment]
It is past time, I'm afraid, but the forum is less important than some of the other items on the backdropcms.org todo list, and without more hands on deck to solve those issues this one will have to wait. Not enough chefs in the kitchen! :(

**Comment by docwilmot** [comment]
O dear, this is on the back burner again is it? 
FWIW I've been pushing this since June 2014, and I do so because I think it *is* as important, or more so. I think, by way of illustration, one of the ways we can grow more "chefs" in the kitchen is to build a more inclusive (easier to get involved with) community. Multiple persons have mentioned that GitHub is not the way to do it, so I think having no good discussion area hurts us.
But I think 2 yrs is enough nagging now, so I'll rest here.

**Comment by LeeteqXV** [comment]
Hi, I am setting up a distributions management community at http://cmx.content.cat , which when ready, we will use for several services, among them, a minipreneurship forum at http://arbolis.net . Currently, at the latter site we have a Drupal Commons setup which we are NOT going to use. However, I wonder if it would be useful if I set up a temporary BackDrop forum discussion group there, to take just these discussions that will lead to decisions about which solution we will aim at. One of the distros we will maintain at cmx.content.cat is exactly a community distro. I think that the BD community would benefit greatly if we could manage to agree on a strategy and specifications for what to build, as the CMX community has at its core aims to set up a funding community to finance further developments of those distros, and we aim at basing our distros on BD core. 

The only requirement for hosting such a temporary group for BD is a security concern: we will require either TOTP (/Google Authenticator-compatible 2FA login) or Yubikey (public server) for each account. This can be up and running within a day or so if you want a forum solution to discuss and explore the options for a forum solution :-)
Since Commons is so far from what we want/need (almost exactly the opposite, given all the lock-in designwise, it is actually useful to explore the bad to make a good spec :-) .

As a side note:
a) given the nature of Social Media and the fact that we can add comments functionality to any node type etc., my personal hunch is that we should regard "forum" as a "view" or "presentation", using the very CM powers that we have at our disposal to make it transparent as a layer on top of any section of the web site. A new/modern "forum" that when desired, will look and feel like a forum, but can work and be presented otherwise simultaneously.
b) there is a "GCC" module developed by the French community, which have not been very willing to think internationally (read: English docs, etc.) but their module might be worth exploring. Seems lightweight and relevant for parts of the wanted functionality. Have not yet tested it much, though. But I will set up a test site with that too, which can be further explored at a subsite of arbolis.net

What we can bring to the table is providing voting options for decision making, organizing content, specs, etc. for various endeavours into these areas. You can have the same copyright of using the content in this group/forum as we have, both can move on in either direction and use all the content elsewhere, for free or premium, without paying us any compensation. Regard the forum as yours and you will be given user roles with permissions to manage the groups/forums. We will first need to find a practical way to manage users, given that some of the modules I would prefer for that part might not work entirely well within Commons. I am about to test that this weekend.

Please let me know if this might be interesting. 
And do have a look at CMX.content.cat - you are among the beneficiaries for its success :-)
Or should I say "we", as I regard myself now as a new/fresh contributor also within BD :-)

By the way; what are you normally using for abbreviation: BD or BDC? My fingers want to type BDC for some reason... (marketingwise, I think it is useful to get newcomers to "see" the CMS part of BD(C) as often as possible; we are going to compete also with WP and Joomla, so BDC might be "smarter"?)

**Comment by jenlampton** [comment]
Wow this is very interesting!

We refer to Backdrop CMS as "Backdrop" for short. There's no D for the same
reason there no R in Drupal: It's a single word.

I have more thoughts and feelings, but too many to type on my phone. More
soon!

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by LeeteqXV** [comment]
Update: I have tested and prepared for possible (temporary) forum sharing at the Arbolis.net site over the weekend. Check here for a writeup/FYI about that:

http://cmx.content.cat/node/16

So, (first) for the BD forum administrators: If you think this it is suitable/desireable to explore as a temporary solution, you may now get user accounts / logins so you can get to see how this might work (and play around with it) before you make decision. Use the contact form over at CMX.content.cat to send me a) the link to your drupal.org user account and b) a related email account with the @backdropcms.org domain, so that I can somehow "verify" it is from the correct person. (PS. See below: it is YOUR accounts that (mostly?) will need 2FA/TOTP/Yubikey.)

(A precision: In my previous info message here I stated that 2FA is a requirement, but that is perhaps mostly relevant for the BD managers user accounts, which will have more permissions than normal users. IF normal users don't need to upload files or post clickable links, they might not need two-factor authentication to secure those accounts. In order to avoid (responsibility for...) XSS/drive-by attacks, etc., we are requiring 2FA for functionality that opens up such risks.)

**Comment by Dinornis** [comment]
I am glad to see that a Backdrop based forum is in the workings for the backdropcms.org site.

I agree with docwilmot, Forums is quite important and not having one at BD.org is counterproductive.  User forum can bring more people in, make it easier to get involved with and grow the Backdrop community.

As an end user of many forums I consider the following features important:

* **Notifications** - Ability to subscribe to individual topics and/or forums and receive email notifications when topic is updated or new topic is added to a specific forum. It should be easy to subscribe when viewing a topic.
* **Notifications Management** - Users should be able to visit a page (for example in their profile) to view a list of the topics/forums they are subscribed to and manage their subscriptions there. A checkbox should be included in the personal settings panel that allows a user to turn on/off auto-subscription to topics she posts or replies to.
* **Search Forums** - Users should be able to search individual or multiple forums, have an option to limit the search to topic titles or the body.  There should be an option to return the search results as a list of topics or posts.
* **Search Topics** - It should be easy for user to search through topic they are viewing.  This is very useful for topics with many posts spanning multiple pages.
* **Editing Post** - Users should be able to edit their posts.  I know this sounds obvious, but many forums do not include this feature.  At the same time it may be a good idea to include a global settings for administrator to allow/disallow editing as some organisations prefer to turn off editing for various reasons.

**Admin/Moderator Features:**
* Gormartsen's suggestions plus
* Word Filter - The [Wordfilter module](https://www.drupal.org/project/wordfilter) works very well and could be ported to cover this and can be used for the whole site.

**Comment by serundeputy** [comment]
I've written a POC stackoverflow type forum here: https://github.com/serundeputy/qa_forum ; this could easily be moved to contrib and allow people to set up forums on their sites.

There is work to be done; currently voting is only for the main `question` content type as a POC.

there are issues filed in the issue queue for `qa_forum` for many but not all of the ideas listed in this thread.

Anyway if anyone is willing to check it out and or participate in the issue queue https://github.com/serundeputy/qa_forum/issues ; I think we could get it up to par in not too much time/effort.

If it seems like a good way to go I can move it to contrib.

**Comment by jenlampton** [comment]
Wow @serundeputy that project looks fantastic! 

My only initial concern is that it looks like it's becoming one module that's going to end up doing all the things a forum needs (comment, vote, follow, etc) where each of those are discrete features that can be done on a forum, or not.  Each also already has existing modules in Drupal, and it would be better for the Backdrop community for those projects to be ported, than for us to re-invent the wheel and build a custom solution.

I'd like to do Backdrop-native solutions for things like content types and views, the configurations related to architecture, but I'd also prefer to use existing Drupal modules for all the bonus features, wherever possible.

**Comment by jenlampton** [comment]
Update: I have added an item to the weekly meeting agenda next week to start moving on a discussion forum. To that end, I just reread the whole thread and added all the recommended features from the comments into the list in the main post (guessed at organization). Let's use this as our starting point for the meeting next week, and see if we can make a decision on technical direction!

**Comment by Dinornis** [comment]
Just wondering, does MVP stand for Minimum Viable Product?

**Comment by serundeputy** [comment]
yes; you got it!

**Comment by jenlampton** [comment]
I've updated the list in the top post based on our discussion in the post-meeting forum-discussion today.  @docwilmot and @Graham-72 we would value your voices in this discussion too. Would you add a comment here if there are any features you feel we are missing, either in the "Minimum Viable Product" list, or the "nice to have" list?

My next steps are to do a code review of the forum module @docwilmot ported to backdrop, and see what's in there other than a content type, taxonomy term, and views. :)

**Comment by jenlampton** [comment]
Hey all, good news! I've startetd work on [forum.backdrpcms.org](https://forum.backdropcms.org) and we now have [our own issue queue](https://github.com/backdrop-ops/forum.backdropcms.org/issues) where we can continue to work out the details :)

I've split the results of this issue into two meta's over there:
* [[META] build out minimum viable product feature set] (https://github.com/backdrop-ops/forum.backdropcms.org/issues/4)
* [[META] build out other nice-to-have features] (https://github.com/backdrop-ops/forum.backdropcms.org/issues/5)

I'm going to close this issue, but please let's not stop the conversation!  :)
